securestorage2/src/main/java/de/adorsys/android/securestorage2/internal/KeyStoreToolApi23.kt

```diff
@@ -0,0 +1,209 @@
+                    SecureStorageConfig.INSTANCE.ENCRYPTION_KEY_ALIAS,
+                    KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
+                )
+                    .setKeyValidityStart(keyStartDate.time)
+                    .setBlockModes(KeyProperties.BLOCK_MODE_CBC)
+                    .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
+                    .setKeySize(256)
+            }
+
+        keyGenerator.init(keyGenParameterSpecBuilder.build())
+        return keyGenerator.generateKey()
+    }
+
+
+    @RequiresApi(23)
+    internal fun encryptValue(
+        context: Context,
+        keyStoreInstance: KeyStore,
+        cipher: Cipher,
+        key: String,
+        value: String
+    ): String {
+
+        val secretKey = if (!keyExists(keyStoreInstance)) {
+            generateKey()
+        } else {
+            getSecretKey(keyStoreInstance)
+        }
+
+        cipher.init(Cipher.ENCRYPT_MODE, secretKey)
+        val bytes = cipher.doFinal(value.toByteArray())
+        saveIVInSecureStorage(context, key, cipher.iv)
+        return Base64.encodeToString(bytes, Base64.DEFAULT)
+    }
+
+    @RequiresApi(23)
+    internal fun decryptValue(
+        context: Context,
+        keyStoreInstance: KeyStore,
+        cipher: Cipher,
+        key: String,
+        value: String
+    ): String {
+
+        val secretKey = getSecretKey(keyStoreInstance)
+
+        cipher.init(
+            Cipher.DECRYPT_MODE, secretKey,
+            IvParameterSpec(getIVFromSecureStorage(context, key))
+        )
+        val encryptedData = Base64.decode(value, Base64.DEFAULT)
+        val decodedData = cipher.doFinal(encryptedData)
+        return String(decodedData)
+    }
+
+    @RequiresApi(23)
+    internal fun deleteKey(context: Context, keyStoreInstance: KeyStore) {
+        // Delete Key from Keystore
+        if (keyExists(keyStoreInstance)) {
+            try {
+                keyStoreInstance.deleteEntry(SecureStorageConfig.INSTANCE.ENCRYPTION_KEY_ALIAS)
+            } catch (e: KeyStoreException) {
+                throw SecureStorageException(
+                    e.message!!,
+                    e,
+                    SecureStorageException.ExceptionType.KEYSTORE_EXCEPTION
+                )
+            }
+        } else {
+            throw SecureStorageException(
+                context.getString(R.string.message_key_does_not_exist),
+                null,
+                SecureStorageException.ExceptionType.KEYSTORE_EXCEPTION
+            )
+        }
+    }
+
+    @RequiresApi(23)
+    private fun saveIVInSecureStorage(context: Context, key: String, iv: ByteArray) {
+        val preferences =
+            context.getSharedPreferences(SecureStorageConfig.INSTANCE.SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE)
+        val encodedIv = Base64.encodeToString(iv, Base64.DEFAULT)
+        preferences.edit().putString("$KEY_CIPHER_IV$key", encodedIv).apply()
```

Comment:
that's weird. Why do you do that? Use `$`, I mean? --> also put into a global variable in a companion object